### PR TITLE
Lift a few limitations on ConstProp optimization pass

### PR DESCRIPTION
**Pull request by cjgillot**
`ConstProp` does not emit lints any more. Therefore, we can extend and simplify it.

This PR starts by fortifying the current implementation of `ConstProp` and of the interpreter.

In a second part, we enable propagation even if overflow checks are disabled. It's now codegen's responsibility to skip emitting overflow checks.

In a third part, we remove limitation on propagation.

Finally, we only disable the pass for mir-opt-level=1, since it's now purely an optimization pass.

### CODE REVIEW

File: compiler/rustc_const_eval/src/interpret/operand.rs [hunk 1]
```diff
@@ -601,7 +601,9 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
         // that the `RevealAll` pass has happened and that the body's consts
         // are normalized, so any call to resolve before that needs to be
         // manually normalized.
-        let val = self.tcx.normalize_erasing_regions(self.param_env, *val);
+        let Ok(val) = self.tcx.try_normalize_erasing_regions(self.param_env, *val) else {
+            throw_unsup!(Normalization)
```
Comment (RalfJung) [inline compiler/rustc_const_eval/src/interpret/operand.rs:605]: This is more invalid than unsupported I think.

File: compiler/rustc_const_eval/src/interpret/cast.rs [hunk 2]
```diff
@@ -356,7 +356,7 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
             }
 
             _ => {
-                span_bug!(self.cur_span(), "invalid unsizing {:?} -> {:?}", src.layout.ty, cast_ty)
+                throw_inval!(Unsize(src.layout.ty, cast_ty))
```
Comment (RalfJung) [inline compiler/rustc_const_eval/src/interpret/cast.rs:359]: Even more const-prop induced technical debt in the interpreter. :(

How can this one even happen...?

**Comment by oli-obk** [comment]
I don't think we should make const prop smarter. We should get the dataflow const prop to cover most of the const prop optimizations and then just remove const prop.

**Comment by saethlin** [comment]
Do we have a solution to DataFlowConstProp's runtime?

File: compiler/rustc_const_eval/src/interpret/cast.rs (hunk 2 above)
Comment (cjgillot) [reply compiler/rustc_const_eval/src/interpret/cast.rs:359]: We can have a generic context
```rust
fn foo<T: ?Sized, V: std::marker::Unsize<T>>(value: &V) -> &T { value }
```

File: compiler/rustc_const_eval/src/interpret/operand.rs (hunk 1 above)
Comment (cjgillot) [reply compiler/rustc_const_eval/src/interpret/operand.rs:605]: `throw_inval!(TooGeneric)` may be better suited then.

**Comment by lcnr** [comment]
r? @oli-obk @RalfJung 

though re 7bbc67468dae49930263f44eb9fedd275acb912e, is there a test which actually ICEs without this change? Because while I can imagine cases where this ICEs with `generic_const_exprs`, that's a bug with that feature and in my ideal impl of CTFE this is actually infallible

**Comment by RalfJung** [comment]
> I don't think we should make const prop smarter. We should get the dataflow const prop to cover most of the const prop optimizations and then just remove const prop.

I think that would be great, if it can be done in our perf budget. It means less code to maintain, and fewer issues with ConstProp using the interpreter in ways it was not intended to be used.

File: compiler/rustc_const_eval/src/interpret/cast.rs (hunk 2 above)
Comment (RalfJung) [reply compiler/rustc_const_eval/src/interpret/cast.rs:359]: Please add a comment explaining that (and that this is only needed for ConstProp).

Why does that not affect the similar `bug!` in unsize_into?

File: compiler/rustc_const_eval/src/interpret/operand.rs (hunk 1 above)
Comment (RalfJung) [reply compiler/rustc_const_eval/src/interpret/operand.rs:605]: Ah right, we already have that variant. Agreed.

**Comment by oli-obk** [comment]
I'll benchmark removing the old pass and enabling the new one so we can make a decision here.

Even if we can't split it, we can still change const prop to stop using mir interpreter rvalue processing and instead make it only use the math APIs like the dataflow const prop does

**Comment by cjgillot** [comment]
On the perf report:
- check builds: mostly non-relevant red and green, nothing obvious;
- debug builds: mostly green, effect of not running ConstProp any more. An exception: keccak (20% regression), as generated MIR is 40% larger.
- opt builds: mostly red. Largest (diesel) may be due to evaluation of ty unevaluated constants (to be confirmed).

**Comment by cjgillot** [comment]
The last perf result confirm there is nothing to gain by extending this pass.
I'll close this PR, and propose the cleanups in another PR based on https://github.com/rust-lang/rust/pull/107921.

**Pull request closed without merging**